### ISSUE TITLE
Feature request (ld65)

### ISSUE THREAD

**Issue by JackAsser1337** [cc65/cc65#648]
When exporting labels using -Ln I would like to have the labels bank byte (if any) set as the upper 8 bits and extend then output from 24bit to 32bit. F.e. a label at offset $dead in bank $12 would be emitted as:

al 1200dead .theLabel

This would later allow VICE to be patched to be bank aware for various carts and display the labels properly in the monitor depending on which bank is active.

**Comment by polluks** [comment]
An extra option? How about expanding -Ln only for bank>0?

**Comment by JackAsser1337** [comment]
Didn’t want to break old builds. There might be cart projects for c64 etc using banks and label output. The extra bank byte in the label output is a breaking changes unless you also have a patched VICE.

**Comment by JackAsser1337** [comment]
I'll clone the repo and fix my stuff. I just gonna check why indirect zp variables aren't auto imported (another problem).

**Comment by oliverschmidt** [comment]
> I'll clone the repo and fix my stuff.

:-)

> Didn’t want to break old builds.

I'm in general rather looking forward than looking backward - I consider it more important to improve things then to be compatible with things.

Please elaborate a bit more on the scenarios which would cause incompatibility and the type of incompatibility.

I seem to understand that your change only makes sense with a changed VICE. Do you have plans on making that VICE change publicly available? If yes, how do those plans look like? Are you in some contact with VICE maintainers?

**Comment by JackAsser1337** [comment]
In that case, as the comments tell in the source, we should completely ditch that output format and instead export a labels file in a more structured format. Then VICE users should transform that output to fit their needs.

The current -Ln options is tailored for the VICE monitor and nothing else as it is today, hence I didn’t want to break it since my change requires a change in VICE also (An unpatched VICE will not accept the output of my -Lb switch at all).

/A

**Comment by polluks** [comment]
@oliverschmidt do you @mrdudz  other hobby? ;-)

**Comment by polluks** [comment]
@JackAsser1337 Oricutron also reads these labels but I don't see breaking compatibility. The Atmos does not use banking beside DOS.

**Comment by JackAsser1337** [comment]
@polluks Adding the bank byte as bit 24-31 in the emitted label is ad-hoc on my part and at least VICE can't handle it. Technically the bank byte should really be a separate field (breaking compatibility even more). Give two emulators now use the current format I'd say it has become a de-facto standard speaking in favour of having my fix as a separate option.

Either that, or completely ditch the current format as it is and have labels emitted with clear bank bytes.

**Comment by polluks** [comment]
You see my trick?
bank 0: 00dead
bank 1: 0100dead

**Comment by JackAsser1337** [comment]
Yes, and a user using the bank-feature and the label-feature and VICE will get broken with your suggestion.

**Comment by oliverschmidt** [comment]
> @oliverschmidt do you know @mrdudz other hobby? ;-)

Sorry but I have no clue what you're on about :-(

**Comment by polluks** [comment]
He is a VICE maintainer...

**Comment by polluks** [comment]
@JackAsser1337 Indeed? Please tell me a scenario.
IMHO you just get an implicit bank command, right?
http://vice-emu.sourceforge.net/vice_12.html#SEC293

**Comment by JackAsser1337** [comment]
polluks: When you import the label file VICE will not be able to parse it. It has NOTHING to do with the bank command in the monitor. The bank command in the monitor is completely unaware of any bank bytes. What a bank is and how big it is, where it is mapped etc etc is completely (almost) up to the cartridge inserted.

**Comment by oliverschmidt** [comment]
Thanks for the hint !

**Comment by polluks** [comment]
Maybe we talk past one another, VICE have to be able to parse 32-bit anyway. I'm just talking about the cc65 part, one option should be sufficient.

**Comment by polluks** [comment]
@oliverschmidt The 8-bit community is not so big. I googled for "prodos 2.5", guess which person I found? ;-)

**Comment by polluks** [comment]
Ok, we are not talking about breaking makefiles of many projects because this change is only used interactive. BTW you may add in ld65.sgml "labels with bank info require VICE 3.2+" :-)

**Comment by mrdudz** [comment]
as for VICE, i'd prefer a format that uses some kind of delimiter between bank and address (and then indeed having a seperate option in ld65 would be the way to go). we can talk about it in a week or so, when 3.2 is out :)

as for banking support in general, one feature i'd really like to see would be a .bank() pseudoop in the assembler. this could return a "bank" value defined in the linker config (per memory area) which then would allow to build proper address tables including the bank value (which could even be the value you need to put into some banking register). it would make some things a LOT easier :)

**Comment by JackAsser1337** [comment]
The .bank() directive already exists! ;) which works as u describe. Each symbol registered in a segment is tagged with the bank id from the corresponding memory area. .bank or ^ operator gives u access to this value. This is also exported and I pick that up when emitting the labels.

/A

**Comment by JackAsser1337** [comment]
@mrdudz I'm f.e. coding a very big project now using EasyFlash and have code scattered through the whole 1Mb ROM. I jsr all over the place between banks automagically by having a jsrf (jsr far) macro. This macro is a tradeoff between complexity, speed, what registers to keep intact and reentrability. So I hade reentrability high on my wish list (i.e. no static regs, zp-location or self-mod code to safe registers, hence I had to use the stack, and ONLY the stack). It's implemented like this:
```
.macro jsrf dest
.scope
	lda #>(return-1)
	pha
	lda #<(return-1)
	pha
	lda BANK
	pha
	lda #>(jsrf_impl-1)
	pha
	lda #<(jsrf_impl-1)
	pha
	lda #>(dest-1)
	pha
	lda #<(dest-1)
	pha
	lda #<.bank(dest)
	pha
	jmp jsrf_impl
	return:
.endscope
.endmacro
```

The jsrf_impl is located in RAM and is implemented like this:
```
.proc jsrf_impl
	pla
	sta BANK
	sta $de00
	rts	; jumps to dest first, then to real return address on the second run
.endproc
```


How it works:

1) Push final return address-1 on the stack. (just like a regular RTS)
2) Push the current BANK on the stack (shadow of write-only EF-register $de00)
3) Push the call adress-1 to the jsrf-impl on the stack.
4) Push the actual destination adress-1 on the stack.
5) Push the bank# for the destination adress on the stack.
6) jmp jsrf_impl
7) pop the BANK index and store to BANK and then to $de00 (important with order here should an IRQ occur)
8) RTS. This will pop the destination address and jump there
9) <Destination runs here>
10) Destination will RTS and thus pop the adress to jsrf_impl and jump there
11) jsrf_impl will pop the old bank value and store it to BANK and then to $de00
12) jsrf_impl will RTS again to the return value pushed first
13) Done
14) PROFIT!

A-reg and flags non-preserved unfortunately. But that's okay for me.

With this scheme I don't have to think about bank switching at all. Some graphics assets will have to be banked switched in manually (bitblt-routined in RAM). But for code it's completely automagical.

**Comment by mrdudz** [comment]
yup, looks like what i had in mind :) so PD will be an EF release? :)
